tests/tracing/test_otel_logging.py

```diff
@@ -0,0 +1,223 @@
+    Without artifact upload, traces won't be retrievable via search_traces.
+    """
+    mlflow.set_tracking_uri(mlflow_server)
+
+    experiment = mlflow.set_experiment("otel-test-experiment")
+    experiment_id = experiment.experiment_id
+
+    resource = OTelSDKResource.create({"service.name": "test-service-e2e"})
+    tracer_provider = TracerProvider(resource=resource)
+
+    exporter = OTLPSpanExporter(
+        endpoint=f"{mlflow_server}/v1/traces", headers={MLFLOW_EXPERIMENT_ID_HEADER: experiment_id}
+    )
+
+    span_processor = BatchSpanProcessor(exporter)
+    tracer_provider.add_span_processor(span_processor)
+
+    otel_trace.set_tracer_provider(tracer_provider)
+    tracer = otel_trace.get_tracer(__name__)
+
+    with tracer.start_as_current_span("otel-e2e-test-span") as span:
+        span.set_attribute("test.e2e.attribute", "e2e-test-value")
+        # Capture the OTel trace ID to verify it matches the MLflow trace ID
+        otel_trace_id = span.get_span_context().trace_id
+
+    flush_success = span_processor.force_flush(10000)
+    assert flush_success, "Failed to flush spans to the server"
+
+    # Wait up to 30 seconds for search_traces() to return a trace
+    traces = []
+    for _ in range(30):
+        traces = mlflow.search_traces(
+            experiment_ids=[experiment_id], include_spans=False, return_type="list"
+        )
+        if traces:
+            break
+        time.sleep(1)
+
+    assert len(traces) > 0, "No traces found in the database after sending spans"
+
+    # Verify the trace ID matches the expected format based on the OTel span
+    expected_trace_id = f"tr-{encode_trace_id(otel_trace_id)}"
+    actual_trace_id = traces[0].info.trace_id
+    assert actual_trace_id == expected_trace_id, (
+        f"Trace ID mismatch: expected {expected_trace_id}, got {actual_trace_id}"
+    )
+
+
+def test_otel_endpoint_requires_experiment_id_header(mlflow_server):
+    """
+    Test that the OTel endpoint requires experiment ID header.
+    """
+    # Create protobuf request
+    span = OTelProtoSpan()
+    span.trace_id = bytes.fromhex("0000000000000002" + "0" * 16)
+    span.span_id = bytes.fromhex("00000002" + "0" * 8)
+    span.name = "test-span-no-header"
+
+    scope = InstrumentationScope()
+    scope.name = "test-scope"
+
+    scope_spans = ScopeSpans()
+    scope_spans.scope.CopyFrom(scope)
+    scope_spans.spans.append(span)
+
+    resource = Resource()
+    resource_spans = ResourceSpans()
+    resource_spans.resource.CopyFrom(resource)
+    resource_spans.scope_spans.append(scope_spans)
+
+    request = ExportTraceServiceRequest()
+    request.resource_spans.append(resource_spans)
+
+    response = requests.post(
+        f"{mlflow_server}/v1/traces",
+        data=request.SerializeToString(),
+        headers={"Content-Type": "application/x-protobuf"},
+        timeout=10,
+    )
+
+    assert response.status_code == 422
+
+
+def test_invalid_otel_span_format_returns_400(mlflow_server):
```

Comment:
```suggestion
def test_invalid_otel_span_format_returns_400(mlflow_server: str):
```

not a blocker, but can we add type hints?